micronota/db/tigrfam.py

```diff
@@ -0,0 +1,266 @@
+        print('Database %s already exists. Skip it.' % prefix)
+        return
+
+    metadata_out = join(out_d, '%s.db' % prefix)
+
+    try:
+        temp_dir = mkdtemp()
+        hmm_tmp = join(temp_dir, hmm)
+        metadata_tmp = join(temp_dir, metadata)
+        with FTP(server) as ftp:
+            ftp.login()
+            ftp.cwd(path)
+
+            # fetch metadata file
+            ftp.retrbinary('RETR %s' % metadata,
+                           open(metadata_tmp, 'wb').write)
+            with tarfile.open(metadata_tmp) as tar:
+                tar.extractall(temp_dir)
+                prepare_metadata(temp_dir, metadata_out)
+
+            # fetch HMM model file
+            ftp.retrbinary('RETR %s' % hmm,
+                           open(hmm_tmp, 'wb').write)
+            # gunzip and move the file
+            with gzip.open(hmm_tmp, 'rb') as i_f, open(hmm_out, 'wb') as o_f:
+                shutil.copyfileobj(i_f, o_f)
+            # don't forget to compress the hmm model file
+            hmmpress_hmm(hmm_out)
+    finally:
+        shutil.rmtree(temp_dir)
+
+
+def prepare_metadata(in_d, fp):
+    '''Compile the metadata into sqlite3 database.
+
+    Parameters
+    ----------
+    in_d : str
+        The input directory containing XXX.INFO files.
+    fp : str
+        The output file path of sqlite3 database.
+
+    Notes
+    -----
+    The schema of the database file contains one table named `tigrfam` that
+    has following columns:
+
+    1. ``id``. TEXT. TIGRFAM ID.
+
+    2. ``tag``. TEXT. 2-letter tag described in the table above in this module.
+
+    3. ``value``. BLOB. The value of the tag.
+
+    4. ``transfer``. INTEGER. Used as the boolean. ``1`` means the ``value``
+       should be transferred to the query sequences as its annotation;
+       ``0`` means not.
+    '''
+    with connect(fp) as conn:
+        conn.execute("DROP TABLE IF EXISTS tigrfam")
+        conn.execute('''CREATE TABLE tigrfam (
+                            id       TEXT    NOT NULL,
+                            tag      TEXT    NOT NULL,
+                            value    BLOB    NOT NULL,
+                            transfer BOOLEAN NOT NULL,
+                        CHECK (transfer IN (0,1)));''')
+
+        for f in os.listdir(in_d):
+            if not f.endswith('.INFO'):
+                continue
+            tigrfam_id = f.split('.', 1)[0]
+            for x in _read_info(join(in_d, f)):
+                conn.execute('''INSERT INTO tigrfam (id, tag, value, transfer)
+                                    VALUES ("%s",?,?,?);''' % tigrfam_id,
+                             x)
+        # don't forget to index the column to speed up query
+        conn.execute('CREATE INDEX id ON tigrfam (id);')
+        conn.commit()
+
+
+def _read_info(fn):
+    '''Parse the .INFO file.
+
+    Parameters
+    ----------
+    fn : str
+        file path
+
+    Yields
+    ------
+    tuple
+        tag, value, int of 0 or 1.
```

Comment:
it is private func. and it is explained at prepare_metadata

On Sun, Jan 31, 2016 at 12:59 PM, mortonjt <notifications@github.com> wrote:

> In micronota/db/tigrfam.py
> <https://github.com/biocore/micronota/pull/40#discussion_r51367284>:
>
> > +        conn.execute('CREATE INDEX id ON tigrfam (id);')
> > +        conn.commit()
> > +
> > +
> > +def _read_info(fn):
> > +    '''Parse the .INFO file.
> > +
> > +    Parameters
> > +    ----------
> > +    fn : str
> > +        file path
> > +
> > +    Yields
> > +    ------
> > +    tuple
> > +        tag, value, int of 0 or 1.
>
> Can you elaborate what the boolean here indicates?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/micronota/pull/40/files#r51367284>.
>